### PR TITLE
[MOD-8198] Introduce INT8 distance functions

### PR DESCRIPTION
**Pull request by meiravgri**
This PR introduces **naive and optimized** distance calculation functions for vectors of `INT8` type elements, for all distance metrics.
Since `INT8` vectors store integral values, we can't normalize these vectors **in place**.
Because of this, unlike other types, `INT8` requires a dedicated implementation for **cosine distance**.

⚠️  The cosine implementation assumes the **norm** is stored at the end of the vector in a `float` number.

#### Requirments
Optimization is available for CPUs and compilers supporting all of the following flags:
* AVX512_F
* AVX512_BW
* AVX512_VL
* AVX512_VNNI

#### Supported dimensions
Optimization is available for vector dimensions **larger than 32**.
Vectors with smaller dimensions will be calculated using naive implementation. 

#### IP and L2
The vectors memory is aligned to memory addresses divisible by `32 * sizeof(int8)` since In each step we load 32-int8 elements
#### Cosine
**We skip aligning the vector's memory address for int8 vectors with cosine distance** due to the extra float stored at the end for the norm.
This changes the alignment requirement to `(dim + sizeof(float)) % 32` for the vector's address to be aligned to 256 bits.
Since vectors that meet this condition necessarily have a residual (since dim % 32 ≠ 0), subsequent loads will be offset during distance calculation.
To avoid adding complexity to the distance function, we do not align the vectors in this case, 
assuming the performance impact is negligible.

## Changes in tests
* changed unit/test_utils.* files name to unit/unit_test_utils.* 
* added general tests_utils file to tests/utils

### CODE REVIEW

**Review by alonre24 — COMMENTED** [review]
:muscle:
Few small comments/questions

File: src/VecSim/spaces/IP/IP_AVX512F_BW_VL_VNNI_INT8.h [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    // We dealt with the residual part. We are left with some multiple of 64-int_8.
+    while (pVect1 < pEnd1) {
+        InnerProductStep(pVect1, pVect2, sum);
+        InnerProductStep(pVect1, pVect2, sum);
+    }
+
+    return _mm512_reduce_add_epi32(sum);
+}
+
+template <unsigned char residual> // 0..32
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/spaces/IP/IP_AVX512F_BW_VL_VNNI_INT8.h:62]: it's not, leftover comment
should be 64

File: src/VecSim/spaces/IP_space.cpp [hunk 2]
```diff
@@ -196,4 +197,56 @@ dist_func_t<float> IP_FP16_GetDistFunc(size_t dim, unsigned char *alignment, con
...
+        if (dim % 32 == 0) // no point in aligning if we have an offsetting residual
+            *alignment = 32 * sizeof(int8_t); // align to 256 bits.
+        return Choose_INT8_IP_implementation_AVX512F_BW_VL_VNNI(dim);
+    }
+#endif
+#endif // __x86_64__
+    return ret_dist_func;
+}
+
+dist_func_t<float> Cosine_INT8_GetDistFunc(size_t dim, unsigned char *alignment,
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/spaces/IP_space.cpp:226]: The only common logic is the alignment intizliation, which we duplicate in all the functions in this file, how is this function different?

[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/spaces/IP_space.cpp:226]: btw note to self: check coverage

File: src/VecSim/spaces/L2/L2_AVX512F_BW_VL_VNNI_INT8.h [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    __m256i temp_b = _mm256_loadu_epi8(pVect2);
+    __m512i vb = _mm512_cvtepi8_epi16(temp_b);
+    pVect2 += 32;
+
+    __m512i diff = _mm512_sub_epi16(va, vb);
+    // _mm512_dpwssd_epi32(src, a, b)
+    // Multiply groups of 2 adjacent pairs of signed 16-bit integers in `a` with corresponding
+    // 16-bit integers in `b`, producing 2 intermediate signed 32-bit results. Sum these 2 results
+    // with the corresponding 32-bit integer in src, and store the packed 32-bit results in dst.
+    sum = _mm512_dpwssd_epi32(sum, diff, diff);
```
Comment (GuyAv46) [inline src/VecSim/spaces/L2/L2_AVX512F_BW_VL_VNNI_INT8.h:23]: Why not use `_mm512_dpbusd_epi32`, avoiding loading only 32 elements and converting them to int16?

File: src/VecSim/spaces/L2/L2_AVX512F_BW_VL_VNNI_INT8.h [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    int8_t *pVect2 = (int8_t *)pVect2v;
+
+    const int8_t *pEnd1 = pVect1 + dimension;
+
+    __m512i sum = _mm512_setzero_epi32();
+
+    // Deal with remainder first. `dim` is more than 32, so we have at least one 32-int_8 block,
+    // so mask loading is guaranteed to be safe
+    if constexpr (residual % 32) {
+        __mmask32 mask = (1LU << (residual % 32)) - 1;
```
Comment (GuyAv46) [inline src/VecSim/spaces/L2/L2_AVX512F_BW_VL_VNNI_INT8.h:39]:
```suggestion
        constexpr __mmask32 mask = (1LU << (residual % 32)) - 1;
```

File: src/VecSim/spaces/IP_space.cpp [hunk 5]
```diff
@@ -196,4 +197,56 @@ dist_func_t<float> IP_FP16_GetDistFunc(size_t dim, unsigned char *alignment, con
...
+        return ret_dist_func;
+    }
+#ifdef CPU_FEATURES_ARCH_X86_64
+    auto features = (arch_opt == nullptr)
+                        ? cpu_features::GetX86Info().features
+                        : *static_cast<const cpu_features::X86Features *>(arch_opt);
+#ifdef OPT_AVX512_F_BW_VL_VNNI
+    if (features.avx512f && features.avx512bw && features.avx512vl && features.avx512vnni) {
+        if (dim % 32 == 0) // no point in aligning if we have an offsetting residual
+            *alignment = 32 * sizeof(int8_t); // align to 256 bits.
```
Comment (GuyAv46) [inline src/VecSim/spaces/IP_space.cpp:245]: alignment should address the additional memory we use for the vector norm

File: src/VecSim/spaces/IP/IP_AVX512F_BW_VL_VNNI_INT8.h [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+    __m256i temp_b = _mm256_loadu_epi8(pVect2);
+    __m512i vb = _mm512_cvtepi8_epi16(temp_b);
+    pVect2 += 32;
+
+    // _mm512_dpwssd_epi32(src, a, b)
+    // Multiply groups of 2 adjacent pairs of signed 16-bit integers in `a` with corresponding
+    // 16-bit integers in `b`, producing 2 intermediate signed 32-bit results. Sum these 2 results
+    // with the corresponding 32-bit integer in src, and store the packed 32-bit results in dst.
+    sum = _mm512_dpwssd_epi32(sum, va, vb);
```
Comment (GuyAv46) [inline src/VecSim/spaces/IP/IP_AVX512F_BW_VL_VNNI_INT8.h:22]: Why not use `_mm512_dpbusd_epi32`, avoiding loading only 32 elements and converting them to int16?

File: src/VecSim/spaces/IP/IP_AVX512F_BW_VL_VNNI_INT8.h [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+float INT8_InnerProductSIMD64_AVX512F_BW_VL_VNNI(const void *pVect1v, const void *pVect2v,
+                                                 size_t dimension) {
+
+    return 1 - INT8_InnerProductImp<residual>(pVect1v, pVect2v, dimension);
+}
+template <unsigned char residual> // 0..32
+float INT8_CosineSIMD64_AVX512F_BW_VL_VNNI(const void *pVect1v, const void *pVect2v,
+                                           size_t dimension) {
+    float norm_v1 = *(float *)((int8_t *)pVect1v + dimension);
+    float norm_v2 = *(float *)((int8_t *)pVect2v + dimension);
```
Comment (GuyAv46) [inline src/VecSim/spaces/IP/IP_AVX512F_BW_VL_VNNI_INT8.h:72]: consider keeping const when casting (and using C++ casting)

File: src/VecSim/spaces/IP/IP.cpp [hunk 8]
```diff
@@ -66,3 +66,26 @@ float FP16_InnerProduct(const void *pVect1, const void *pVect2, size_t dimension
...
+    return res;
+}
+
+float INT8_InnerProduct(const void *pVect1v, const void *pVect2v, size_t dimension) {
+    return 1 - INT8_InnerProductImp(pVect1v, pVect2v, dimension);
+}
+
+float INT8_Cosine(const void *pVect1v, const void *pVect2v, size_t dimension) {
+    float norm_v1 = *(float *)((int8_t *)pVect1v + dimension);
+    float norm_v2 = *(float *)((int8_t *)pVect2v + dimension);
```
Comment (GuyAv46) [inline src/VecSim/spaces/IP/IP.cpp:89]: consider keeping const when casting (and using C++ casting)

File: src/VecSim/spaces/IP/IP_AVX512F_BW_VL_VNNI_INT8.h [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    return 1.0f -
+           float(INT8_InnerProductImp<residual>(pVect1v, pVect2v, dimension)) / (norm_v1 * norm_v2);
```
Comment (GuyAv46) [inline src/VecSim/spaces/IP/IP_AVX512F_BW_VL_VNNI_INT8.h:74]: have you tried first calculating the `INT8_InnerProductImp` result and then "loading" the norms? worried about cache locality (when reading the norm from the end of the vector)

Comment (meiravgri) [reply src/VecSim/spaces/IP/IP_AVX512F_BW_VL_VNNI_INT8.h:74]: ill switch

File: tests/unit/test_spaces.cpp [hunk 10]
```diff
@@ -211,6 +228,36 @@ TEST_F(SpacesTest, fp16_ip_no_optimization_func_test) {
...
+
+    test_utils::populate_int8_vec(v1, dim, 123);
+    test_utils::populate_int8_vec(v2, dim, 123);
+
+    // write the norm at the end of the vector
+    *(float *)(v1 + dim) = test_utils::compute_norm(v1, dim);
+    *(float *)(v2 + dim) = test_utils::compute_norm(v2, dim);
+
+    float dist = INT8_Cosine((const void *)v1, (const void *)v2, dim);
+    ASSERT_NEAR(dist, 0.0, 0.000001);
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply tests/unit/test_spaces.cpp:256]:
```
Expected equality of these values:
  dist
    Which is: 1.1920929e-07
  0.0f
    Which is: 0
```

File: src/VecSim/spaces/IP/IP_AVX512F_BW_VL_VNNI_INT8.h (hunk 9 above)
Comment (meiravgri) [reply src/VecSim/spaces/IP/IP_AVX512F_BW_VL_VNNI_INT8.h:74]: ill switch

File: src/VecSim/spaces/IP/IP_AVX512F_BW_VL_VNNI_INT8.h (hunk 6 above)
Comment (meiravgri) [reply src/VecSim/spaces/IP/IP_AVX512F_BW_VL_VNNI_INT8.h:22]: nice catch!
Unfortunately, `_mm512_dpbusd_epi32` expects `a` to be unsigned 8-bit integers vecor, and `b` to be signed 8-bit vector
> Multiply groups of 4 adjacent pairs of unsigned 8-bit integers in a with corresponding signed 8-bit integers in b

So although indeed faster, it gives the wrong results 

```
Benchmark                                  Time             CPU   Iterations UserCounters...
--------------------------------------------------------------------------------------------
IP_dpwssd/Dim:512                       5.41 ns         5.41 ns    128425185 dist=-1.024k
IP_dpbusd/Dim:512                       3.26 ns         3.26 ns    218138693 dist=130.048k
IP_sanity_bm/Dim:512/iterations:1        490 ns          360 ns            1 dist=-1.024k
```

Comment (GuyAv46) [reply src/VecSim/spaces/IP/IP_AVX512F_BW_VL_VNNI_INT8.h:22]: Bummer... So there is no function for multiplying int8*int8 or uint8*uint8? Only an unsigned by a signed?

File: src/VecSim/spaces/IP_space.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/spaces/IP_space.cpp:226]: coverage: adding cosine to small dimension spaces unit tests

File: src/VecSim/spaces/IP/IP_AVX512F_BW_VL_VNNI_INT8.h (hunk 6 above)
Comment (meiravgri) [reply src/VecSim/spaces/IP/IP_AVX512F_BW_VL_VNNI_INT8.h:22]: nope
we might want to consider using `_mm512_unpackhi_epi8` and `_mm512_unpacklo_epi8` for `uint8_t`

**Review by GuyAv46 — APPROVED** [review]
🤯

**Pull request merged**